staticaddr/loopin/manager.go

```diff
@@ -0,0 +1,345 @@
+			"exclusive")
+	}
+
+	// If private is set, we generate route hints.
+	if req.Private {
+		// If last_hop is set, we'll only add channels with peers set to
+		// the last_hop parameter.
+		includeNodes := make(map[route.Vertex]struct{})
+		if req.LastHop != nil {
+			includeNodes[*req.LastHop] = struct{}{}
+		}
+
+		// Because the Private flag is set, we'll generate our own set
+		// of hop hints.
+		req.RouteHints, err = loop.SelectHopHints(
+			m.runCtx, m.cfg.LndClient, totalDepositAmount,
+			loop.DefaultMaxHopHints, includeNodes,
+		)
+		if err != nil {
+			return fmt.Errorf("unable to generate hop hints: %w",
+				err)
+		}
+	}
+
+	// Request current server loop in terms and use these to calculate the
+	// swap fee that we should subtract from the swap amount in the payment
+	// request that we send to the server. We pass nil as optional route
+	// hints as hop hint selection when generating invoices with private
+	// channels is an LND side black box feature. Advanced users will quote
+	// directly anyway and there they have the option to add specific route
+	// hints.
+	// The quote call will also request a probe from the server to ensure
+	// feasibility of a loop-in for the totalDepositAmount.
+	numDeposits := uint32(len(deposits))
+	quote, err := m.cfg.QuoteGetter.GetLoopInQuote(
+		m.runCtx, totalDepositAmount, m.cfg.NodePubkey, req.LastHop,
+		req.RouteHints, req.Initiator, numDeposits,
+	)
+	if err != nil {
+		return fmt.Errorf("unable to get loop in quote: %w", err)
+	}
+
+	// If the previously accepted quote fee is lower than what is quoted now
+	// we abort the swap.
+	if quote.SwapFee > req.MaxSwapFee {
+		log.Warnf("Swap fee %v exceeding maximum of %v",
+			quote.SwapFee, req.MaxSwapFee)
+
+		return loop.ErrSwapFeeTooHigh
+	}
+
+	paymentTimeoutSeconds := uint32(DefaultPaymentTimeoutSeconds)
+	if req.PaymentTimeoutSeconds != 0 {
+		paymentTimeoutSeconds = req.PaymentTimeoutSeconds
+	}
+
+	swap := &StaticAddressLoopIn{
+		DepositOutpoints:      req.DepositOutpoints,
+		Deposits:              deposits,
+		Label:                 req.Label,
+		Initiator:             req.Initiator,
+		RouteHints:            req.RouteHints,
+		QuotedSwapFee:         quote.SwapFee,
+		MaxSwapFee:            req.MaxSwapFee,
+		PaymentTimeoutSeconds: paymentTimeoutSeconds,
+	}
+	if req.LastHop != nil {
+		swap.LastHop = req.LastHop[:]
+	}
+
+	m.Lock()
+	swap.InitiationHeight = m.currentHeight
+	m.Unlock()
+
+	return m.startLoopInFsm(swap)
+}
+
+// startLoopInFsm initiates a loop-in state machine based on the user-provided
+// swap information, sends that info to the server and waits for the server to
+// return htlc signature information. It then creates the loop-in object in the
+// database.
+func (m *Manager) startLoopInFsm(loopIn *StaticAddressLoopIn) error {
+	// Create a state machine for a given deposit.
+	recovery := false
+	loopInFsm, err := NewFSM(m.runCtx, loopIn, m.cfg, recovery)
```

Comment:
Again I think we should store active FSM's in order to avoid accidental GC.

Reply:
done.